src/blob_gc_job.cc

```diff
@@ -222,31 +239,64 @@ Status BlobGCJob::DoRunGC() {
 
     BlobRecord blob_record;
     blob_record.key = gc_iter->key();
+    blob_record.sequence = gc_iter->sequence();
     blob_record.value = gc_iter->value();
     // count written bytes for new blob record,
     // blob index's size is counted in `RewriteValidKeyToLSM`
     metrics_.bytes_written += blob_record.size();
 
-    BlobIndex new_blob_index;
+    MergeBlobIndex new_blob_index;
     new_blob_index.file_number = blob_file_handle->GetNumber();
+    new_blob_index.sequence = gc_iter->sequence();
+    new_blob_index.source_file_number = blob_index.file_number;
     blob_file_builder->Add(blob_record, &new_blob_index.blob_handle);
     std::string index_entry;
-    new_blob_index.EncodeTo(&index_entry);
-
-    // Store WriteBatch for rewriting new Key-Index pairs to LSM
-    GarbageCollectionWriteCallback callback(cfh, blob_record.key.ToString(),
-                                            std::move(blob_index));
-    callback.value = index_entry;
-    rewrite_batches_.emplace_back(
-        std::make_pair(WriteBatch(), std::move(callback)));
-    auto& wb = rewrite_batches_.back().first;
-    s = WriteBatchInternal::PutBlobIndex(&wb, cfh->GetID(), blob_record.key,
-                                         index_entry);
+
+    if (gc_rewrite_mode_ == TitanGcRewriteMode::kDefault) {
+      new_blob_index.EncodeToBase(&index_entry);
+      // Store WriteBatch for rewriting new Key-Index pairs to LSM
+      GarbageCollectionWriteCallback callback(cfh, blob_record.key.ToString(),
+                                              std::move(blob_index));
+      callback.value = index_entry;
+      rewrite_batches_.emplace_back(
+          std::make_pair(WriteBatch(), std::move(callback)));
+      auto& wb = rewrite_batches_.back().first;
+      s = WriteBatchInternal::PutBlobIndex(&wb, cfh->GetID(), blob_record.key,
+                                           index_entry);
+    } else if (gc_rewrite_mode_ == TitanGcRewriteMode::kMerge) {
+      new_blob_index.EncodeTo(&index_entry);
+      rewrite_batches_without_callback_.emplace_back(WriteBatch());
+      auto& wb = rewrite_batches_without_callback_.back();
+      s = WriteBatchInternal::Merge(&wb, cfh->GetID(), blob_record.key,
+                                    index_entry);
+    } else {  // TitanGcRewriteMode::kIngest / TitanGcRewriteMode::kFastIngest
+      new_blob_index.EncodeTo(&index_entry);
+      s = blob_index_table.Merge(blob_record.key, index_entry);
+      if (s.ok()) {
+        entry_count++;
+      }
+    }
     if (!s.ok()) {
       break;
     }
   }
 
+  if (s.ok() && gc_rewrite_mode_ >= TitanGcRewriteMode::kIngest) {
+    if (entry_count > 0) {
+      ingestion_file_status_ = blob_index_table.Finish();
```

Comment:
how about calling it in `RewriteValidKeyToLSM`,  so we can remove `ingestion_file_ready_`